### PR TITLE
fix(connector): [Shift4]Add Refund webhooks 

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Refund webhook support for shift4 ([status mapping ref](https://dev.shift4.com/docs/api#event-types)) . We always update the status by psync as we don't have source verification for shift4 .

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

**payment webhook**
Screenshot 2023-05-29 at 6 51 42 PM
**Refund webhook**
Screenshot 2023-05-29 at 6 51 05 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/shift4.rs [hunk 1]
```diff
@@ -730,9 +730,25 @@ impl api::IncomingWebhook for Shift4 {
...
+                        .data
+                        .refunds
+                        .and_then(|refund| {
+                            refund.first().map(|refund_object| refund_object.id.clone())
+                        })
+                        .ok_or(errors::ConnectorError::WebhookReferenceIdNotFound)?,
+                ),
+            ));
+        }
+        Err(errors::ConnectorError::WebhookReferenceIdNotFound).into_report()
```
Comment (prasunna09) [inline crates/router/src/connector/shift4.rs:751]:
```suggestion
        if shift4::is_transaction_event(&details.event_type) {
            Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
                api_models::payments::PaymentIdType::ConnectorTransactionId(details.data.id),
            ))
        } else if shift4::is_refund_event(&details.event_type) {
            Ok(api_models::webhooks::ObjectReferenceId::RefundId(
                api_models::webhooks::RefundIdType::ConnectorRefundId(
                    details
                        .data
                        .refunds
                        .and_then(|refund| {
                            refund.first().map(|refund_object| refund_object.id.clone())
                        })
                        .ok_or(errors::ConnectorError::WebhookReferenceIdNotFound)?,
                ),
            ))
        } else {
            Err(errors::ConnectorError::WebhookEventTypeNotFound).into_report()
        }
```

Comment (ArjunKarthik) [reply crates/router/src/connector/shift4.rs:751]: @prasunna09 if you requested some changes from the author please use the label waiting on author

**Review by prasunna09 — APPROVED** [review]

File: crates/router/src/connector/shift4/transformers.rs [hunk 2]
```diff
@@ -602,3 +615,30 @@ pub struct ApiErrorResponse {
...
+}
+
+pub fn is_refund_event(event: &Shift4WebhookEvent) -> bool {
+    matches!(event, Shift4WebhookEvent::ChargeRefunded)
+}
+
+impl From<Shift4WebhookEvent> for api::IncomingWebhookEvent {
+    fn from(event: Shift4WebhookEvent) -> Self {
+        match event {
+            Shift4WebhookEvent::ChargeSucceeded | Shift4WebhookEvent::ChargeUpdated => {
```
Comment (jagan-jaya) [inline crates/router/src/connector/shift4/transformers.rs:636]: why ChargeSucceeded is also mapped as `PaymentIntentProcessing`
do we get `ChargeCaptured` for `automatic` capture_method also from shift4?

Comment (srujanchikke) [reply crates/router/src/connector/shift4/transformers.rs:638]: [](https://dev.shift4.com/docs/api#event-types ) ChargeSucceded is when charge is created sucessfully , and charge captured Occurs when a charge is captured. No we are not getting chargeCaptured for automatic but , as per their documentation i have mapped the status .

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**